lxd/storage/drivers/drivers_lvm_utils.go

```diff
@@ -0,0 +1,693 @@
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("Error checking for LVM thin pool \"%s\"", poolName)
+	}
+
+	// Found LV named poolname, check type:
+	attrs := strings.TrimSpace(string(output[:]))
+	if strings.HasPrefix(attrs, "t") {
+		return true, nil
+	}
+
+	return false, fmt.Errorf("LVM volume named \"%s\" exists but is not a thin pool", poolName)
+}
+
+// logicalVolumeExists checks whether the specified logical volume exists.
+func (d *lvm) logicalVolumeExists(volDevPath string) (bool, error) {
+	_, err := shared.RunCommand("lvs", "--noheadings", "-o", "lv_name", volDevPath)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("Error checking for LVM logical volume \"%s\"", volDevPath)
+	}
+
+	return true, nil
+}
+
+// createThinpool creates a thin pool logical volume.
+func (d *lvm) createThinpool(lvmVersion string, vgName string, thinPoolName string) error {
+	exists, err := d.thinpoolExists(vgName, thinPoolName)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		return nil
+	}
+
+	err = d.createDefaultThinPool(lvmVersion, vgName, thinPoolName)
+	if err != nil {
+		return err
+	}
+
+	poolExists, err := d.thinpoolExists(vgName, thinPoolName)
+	if err != nil {
+		return fmt.Errorf("Error checking for LVM thin pool \"%s\" in \"%s\": %v", thinPoolName, vgName, err)
+	}
+
+	if !poolExists {
+		return fmt.Errorf("LVM thin pool \"'%s\" does not exist in Volume Group \"%s\"", thinPoolName, vgName)
+	}
+
+	return nil
+}
+
+// createDefaultThinPool creates the default thinpool as 100% the size of the volume group with a 1G
+// meta data volume.
+func (d *lvm) createDefaultThinPool(lvmVersion, vgName, thinPoolName string) error {
+	isRecent, err := d.lvmVersionIsAtLeast(lvmVersion, "2.02.99")
+	if err != nil {
+		return fmt.Errorf("Error checking LVM version: %s", err)
+	}
+
+	// Create the thin pool
+	lvmThinPool := fmt.Sprintf("%s/%s", vgName, thinPoolName)
+	if isRecent {
+		_, err = shared.TryRunCommand(
+			"lvcreate",
+			"-Wy", "--yes",
+			"--poolmetadatasize", "1G",
+			"-l", "100%FREE",
+			"--thinpool", lvmThinPool)
+	} else {
+		_, err = shared.TryRunCommand(
+			"lvcreate",
+			"-Wy", "--yes",
+			"--poolmetadatasize", "1G",
+			"-L", "1G",
+			"--thinpool", lvmThinPool)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error creating LVM thin pool named %s: %v", thinPoolName, err)
+	}
+
+	if !isRecent {
+		// Grow it to the maximum VG size (two step process required by old LVM)
+		_, err = shared.TryRunCommand("lvextend", "--alloc", "anywhere", "-l", "100%FREE", lvmThinPool)
+
+		if err != nil {
+			return fmt.Errorf("Error growing LVM thin pool named %s: %v", thinPoolName, err)
+		}
+	}
+
+	return nil
+}
+
+// renameThinpool renames a thinpool volume.
+func (d *lvm) renameThinpool(vgName, thinPoolName, newThinPoolName string) error {
```

Comment:
Only has a single caller, any reason to keep it separate?

Reply:
No, I will merge into `Updated`.